### DISCUSSION TITLE
Control binding questions.

### DISCUSSION THREAD

**Discussion by g40** [AvaloniaUI/Avalonia#10787] [General]
On the face of it, a `Button` and a `MenuItem` are quite similar. But wiring them up in XAML is really different. 

`<Button Click="Button_Click">Hello World</Button>` works and calls the Button_Click method on a **Window** derived instance without any further fuss. The handler also has a desirable signature: `public void Button_Click(object sender, RoutedEventArgs e)`. As a big bonus the XAML compilation will fail if the Click handler cannot be resolved. Nice.

In the case of the MenuItem , there is no simple `Click` parameter, and the following will not compile
`<MenuItem Name="FileExit" Header="_Exit" Command="{CompiledBinding OnExitCommand}"/>`, the error being
`Avalonia error XAMLIL: Unable to resolve property of name 'OnExitCommand' on type 'AvaloniaApplication1.MainWindow'`

Fair enough, the documentation here (https://docs.avaloniaui.net/docs/data-binding/compiledbindings) states that "We cannot use compiled bindings to bind to methods, so we use the normal Binding". 

So binding for the `MenuItem` only appears to work if this dynamic form is used:
`MenuItem Name="FileExit" Header="_Exit" Command="{ReflectionBinding $parent[Window].OnExitCommand}"/>`

which is a shame as compile-time checking is sacrificed and performance probably takes a hit too, though for a menu not such a big deal.

Note too that the `Window` indexer appears to be completely undocumented. https://docs.avaloniaui.net/docs/data-binding/binding-to-controls

Questions arising then:

Why does the `Button` have an easy to use, compile-time checked, `Click` parameter and the `MenuItem` does not? 
How easy is it to add `Cllck` to the `MenuItem` class?

Longer term, is there no way to compile time time method bindings? I can see it might be a bit hairy in the case of overrides but the 
`x:DataType="vm:MainWindow"` should take care of that?

Thanks for listening.

**Comment by mysteryx93** [comment]
There is CompileBindings, set this on each control

    x:CompileBindings="True"

or enable it project-wide

    <AvaloniaUseCompiledBindingsByDefault>true</AvaloniaUseCompiledBindingsByDefault>

You then need to set `x:DataType` for all your views

Note that there are still a few bugs with compile bindings, which is why it's not enabled by default.

**Comment by timunie** [comment]
@g40 

- Are you sure the `DataType` is correct? In a normal case it should be `MainViewModel` and not `MainWindow`
- binding to methods will work in 11.0 
- I highly suggest to add commands instead of binding to methods. Here is a sample: https://github.com/AvaloniaUI/Avalonia.Samples/tree/main/src/Avalonia.Samples/MVVM/CommandSample

**Reply by timunie** [reply]
Addition: If you prefer EventHandlers (like Click on a button) instead of Commands, you can also listen to `PointerPressed` or `PointerReleased` which will listen to your mouse buttons clicked or released.
